### PR TITLE
SDKConfigValues, Validations, Uninstall scenario

### PR DESCRIPTION
**Pull request by camilamacedo86**
## Motivation
https://issues.jboss.org/browse/AEROGEAR-9240
https://issues.jboss.org/browse/AEROGEAR-9224

## What
- Change the SDKConfig struct as it was requested
- Move validation of cluster protocol to MSS pkg. It should not be used for another CR
- Fix finalizer for the scenario when the service was deleted (UNINSTALL)
- Rename the const in order to follow the golang convection
 
## Why
- Keep easy to be maintained
- Attend requirements. 

## Verification Steps
Use the image: `docker.io/cmacedo/mobile-security-service-operator:SDK-CONFIG`
1. Install all
2. Create an app and check that SDKConfig is as defined in the trello card. (https://trello.com/c/ExZqTlJ6/12-5-mss-consumption-cr-as-a-developer-i-want-to-be-able-to-specify-necessary-info-through-a-cr-such-that-i-can-associate-my-mobile)
4. Ensure that the app is created in the Service
5. Delete the app and ensure that the app was removed from the service 
6. Test the delete-all and check that it should not be stuck. 

## Checklist:

- [X] Code has been tested locally by PR requester
- [ ] Changes have been successfully verified by another team member 

## Progress

- [x] Finished task
- [ ] TODO

Screenshot 2019-05-23 at 10 32 22
Screenshot 2019-05-23 at 10 32 15

### CODE REVIEW

**Comment by davidffrench** [comment]
Great work @camilamacedo86 . I have reviewed the code and verified it is working as expected. This needs to be rebased and the format of the config needs to be changed.

As discussed in https://trello.com/c/ExZqTlJ6/12-5-mss-consumption-cr-as-a-developer-i-want-to-be-able-to-specify-necessary-info-through-a-cr-such-that-i-can-associate-my-mobile, just the url is enough.

@wei-lee I have a few clarifying questions to make sure we don't need to make any additional changes in the future.
1. In the new RHMDS architecture, does the mobile client SDK still expect there to be a single `mobile-services.json` file with the connection details for all bound services? Or will it be multiple files in the SDK?
2. Does MDC expect the filename to contain some substring to identify which config maps it will consume for an individual app?
3. Does MDC expect there to be any other details in the Config Map generated by each service such as appName or type (e.g. MSS or UPS)?

**Comment by wei-lee** [comment]
@davidffrench 

> In the new RHMDS architecture, does the mobile client SDK still expect there to be a single mobile-services.json file with the connection details for all bound services? Or will it be multiple files in the SDK?

Yes, a single config file.

> Does MDC expect the filename to contain some substring to identify which config maps it will consume for an individual app?

Prefer using labels. E.g. `org.aerogear/mobile-client=appname`. But TBH MDC doesn't really expect anything, nor it does make any assumptions. It's up to the service itself to decide what's the best way to annotate the configmap, as long as it's consistent and easy to understand. 

> Does MDC expect there to be any other details in the Config Map generated by each service such as appName or type (e.g. MSS or UPS)?

No, it doesn't.

**Comment by davidffrench** [comment]
> @davidffrench
> 
> > In the new RHMDS architecture, does the mobile client SDK still expect there to be a single mobile-services.json file with the connection details for all bound services? Or will it be multiple files in the SDK?
> 
> Yes, a single config file.
> 
Thank you

> > Does MDC expect the filename to contain some substring to identify which config maps it will consume for an individual app?
> 
> Prefer using labels. E.g. `org.aerogear/mobile-client=appname`. But TBH MDC doesn't really expect anything, nor it does make any assumptions. It's up to the service itself to decide what's the best way to annotate the configmap, as long as it's consistent and easy to understand.
> 
This makes perfect sense. 

> > Does MDC expect there to be any other details in the Config Map generated by each service such as appName or type (e.g. MSS or UPS)?
> 
> No, it doesn't.

One further question, I was under the assumption the logic in MDC would be generic to consume each separate service bound to a single app and generate the mobile-services.json for that specific app. If each service is allowed to annotate their config maps in their own way and there isn't required to specify a unique identifier in a particular way through annotations or otherwise for a single app, how will MDC know what config maps are associated with which apps and how it should consume a specific services config map to generate the mobile-services.json? We can chat more about this in the office tomorrow if you would prefer.

**Comment by wei-lee** [comment]
@davidffrench 

> One further question, I was under the assumption the logic in MDC would be generic to consume each separate service bound to a single app and generate the mobile-services.json for that specific app. If each service is allowed to annotate their config maps in their own way and there isn't required to specify a unique identifier in a particular way through annotations or otherwise for a single app, how will MDC know what config maps are associated with which apps and how it should consume a specific services config map to generate the mobile-services.json? We can chat more about this in the office tomorrow if you would prefer.

It's not required to have a generic login in MDC to find all the configmaps for a given client app. We only have to support like 4 services anyway so it's not a big deal if some of them use different way to annotation the configmaps (or may not even generate one). The only requirement is the app name is used somewhere in the config map, like be part of a label, or part the of configmap name etc so that MDC can find it using the app name.

**Comment by austincunningham** [comment]
two fields are required to work with the sdk and the ionic showcase

{
  "type": "security",
  "url": "http://route-mobile-security-service.192.168.99.100.nip.io/api"
}
https://github.com/aerogear/ionic-showcase/blob/b23f58f96cee57397a73ee059fd9250825f86fe6/src/app/services/openshift.service.ts#L44

https://github.com/aerogear/aerogear-js-sdk/blob/20a0211a78cdedd50797e7a394f81d3d59fa1367/packages/security/src/appSecurity/AppSecurity.ts#L32

**Comment by davidffrench** [comment]
> @davidffrench
> 
> > One further question, I was under the assumption the logic in MDC would be generic to consume each separate service bound to a single app and generate the mobile-services.json for that specific app. If each service is allowed to annotate their config maps in their own way and there isn't required to specify a unique identifier in a particular way through annotations or otherwise for a single app, how will MDC know what config maps are associated with which apps and how it should consume a specific services config map to generate the mobile-services.json? We can chat more about this in the office tomorrow if you would prefer.
> 
> It's not required to have a generic login in MDC to find all the configmaps for a given client app. We only have to support like 4 services anyway so it's not a big deal if some of them use different way to annotation the configmaps (or may not even generate one). The only requirement is the app name is used somewhere in the config map, like be part of a label, or part the of configmap name etc so that MDC can find it using the app name.

Perfect, thanks @wei-lee . Would you also like us to implement the logic in MDC to consume the configmap or just advise you what annotations are where? We can create a JIRA for this.

**Comment by camilamacedo86** [comment]
HI @davidffrench @wei-lee,

Following how it works in MSS Oper, currently

* One SDKConfigMap is created for each app in the app namespace
* The SDKConfigMap is using the labels as defined
* The name of SDKConfigMap has the name of the app
* SDKConfigMap has an attribute with the name of the app as well. 
* SDKConfigMap has the public URL with the init endpoint

Screenshot 2019-05-22 at 14 33 16

**Conclusion:** I understand that it is attending all the requirements. 
However, @wei-lee , could you check to follow what do would like to be changed here?

- [ ] - SDKConfigMap should be created in the Operator namespace instead of the MDC/App namespace
- [ ] - I do not would like to have all these attributes. I want just the URL
- [ ] - I would like to change the JSON and I want it as "<add JSON and description fields>"
- [ ] - I am OK with as it is now and no changes are required. 
- [ ] - Other: <Please, specify what exactly you would like that be changed>

c/c @davidffrench

File: pkg/models/sdk_config.go [hunk 1]
```diff
@@ -5,28 +5,29 @@ import (
...
 func NewSDKConfig(m *mobilesecurityservicev1alpha1.MobileSecurityServiceApp, serviceUrl string) *SDKConfig {
-
 	cfg := new(SDKConfig)
-	cfg.Version = "1.0.0"
+	cfg.ID = m.Spec.AppId
 	cfg.Name = m.Spec.AppName
-	cfg.ClientID = m.Spec.AppId
-	cfg.Namespace = m.Namespace
-	cfg.Services = getServices(m, serviceUrl)
+	cfg.Type = m.Name
```
Comment (austincunningham) [inline pkg/models/sdk_config.go]: type needs to be "security" to work with the showcase app

File: pkg/models/sdk_config.go [hunk 2]
```diff
@@ -5,28 +5,29 @@ import (
...
+	ID            string        `json:"id"`
+	Name          string        `json:"name"`
+	Type          string        `json:"type"`
+	URL           string        `json:"url"`
+	ConfigService ConfigService `json:"config,omitempty"`
 }
 
 func NewSDKConfig(m *mobilesecurityservicev1alpha1.MobileSecurityServiceApp, serviceUrl string) *SDKConfig {
-
 	cfg := new(SDKConfig)
```
Comment (craicoverflow) [inline pkg/models/sdk_config.go]: Constructing a model like this is not a common practice in Go, and you are taking multiple steps to build the `SDKConfig` struct, instead of 1.

`new(x)` sets each value in the struct to its default value. Next you are overwriting each value individually.

It would be better to define your constructor like so:

```go
func NewSDKConfig(m *mobilesecurityservicev1alpha1.MobileSecurityServiceApp, serviceUrl string) *SDKConfig {
	return &SDKConfig{
		ID:            m.Spec.AppId,
		Name:          m.Spec.AppName,
		Type:          "security",
		URL:           serviceUrl,
		ConfigService: *NewConfigService(serviceUrl),
	}
}
```

**Comment by davidffrench** [comment]
@camilamacedo86 Here are the changes that should be done.

1. Change generated configmap name from `<appName>-sdk` to `<appName>-security` 
2. The annotations should be the following:
  - type: mobilesecurityservice
  - mobilesecurityserviceapp_cr: mobile-security-service-app
  - appname: <appName>
3. The JSON should just be "{ "url": "http://route-mobile-security-service.192.168.99.100.nip.io/api" }"

File: pkg/models/sdk_config.go (hunk 1 above)
Comment (camilamacedo86) [reply pkg/models/sdk_config.go]: Done :-)

**Comment by camilamacedo86** [comment]
Hi @davidffrench,

Done as your last comment. Note that you select the field which says that the JSON should have the type then after that you wrote that note.

File: pkg/controller/mobilesecurityservice/helpers.go [hunk 3]
```diff
@@ -6,10 +6,11 @@ import (
 
 	mobilesecurityservicev1alpha1 "github.com/aerogear/mobile-security-service-operator/pkg/apis/mobilesecurityservice/v1alpha1"
 	"github.com/aerogear/mobile-security-service-operator/pkg/utils"
-	"github.com/go-logr/logr"
 	corev1 "k8s.io/api/core/v1"
 )
 
+const LETTER_BYTES = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ"
```
[the remark this answers is not in the document]
Comment (camilamacedo86) [reply pkg/controller/mobilesecurityservice/helpers.go]: Hi @craicoverflow,

It is a convection use upper case for CONSTANTS. The golang is complaining because of "_". It for this language specifically the `is to use MixedCaps or mixedCaps rather than underscores to write multiword names`.
 
I will do changes for all follow the go convection.

File: pkg/controller/mobilesecurityserviceapp/controller.go [hunk 4]
```diff
@@ -133,25 +133,42 @@ func (r *ReconcileMobileSecurityServiceApp) Reconcile(request reconcile.Request)
...
 
 	//Get the REST Service Endpoint
 	serviceAPI := service.GetServiceAPIURL(mssInstance)
 
 	//Check if the APP CR was marked to be deleted
 	isAppMarkedToBeDeleted := instance.GetDeletionTimestamp() != nil
 	if isAppMarkedToBeDeleted {
+
+		// If the Service was deleted and/or marked to be deleted
+		if err := r.client.Get(context.TODO(), types.NamespacedName{Name: "mobile-security-service", Namespace: operatorNamespace}, mssInstance); err != nil || mssInstance.GetDeletionTimestamp() != nil {
```
[the remark this answers is not in the document]
Comment (camilamacedo86) [reply pkg/controller/mobilesecurityserviceapp/controller.go]: not go of the PR

File: pkg/controller/mobilesecurityservice/validations.go [hunk 5]
```diff
@@ -1,11 +1,22 @@
...
+package mobilesecurityservice
 
 import (
 	mobilesecurityservicev1alpha1 "github.com/aerogear/mobile-security-service-operator/pkg/apis/mobilesecurityservice/v1alpha1"
 	"github.com/go-logr/logr"
 )
 
-func CheckClusterProtocol(serviceInstance *mobilesecurityservicev1alpha1.MobileSecurityService, reqLogger logr.Logger) bool {
+//Check if the mandatory specs are filled
+func hasMandatorySpecs(serviceInstance *mobilesecurityservicev1alpha1.MobileSecurityService, reqLogger logr.Logger) bool {
```
[the remark this answers is not in the document]
Comment (camilamacedo86) [reply pkg/controller/mobilesecurityservice/validations.go]: Not go of the PR can be made after in another one
It was already there.

File: pkg/controller/mobilesecurityservice/validations.go [hunk 6]
```diff
@@ -1,11 +1,22 @@
...
+	//Check the values defined for the ClusterProtocol in the MobileSecurityService CR
+	if res := checkClusterProtocol(serviceInstance, reqLogger); res != true {
```
[the remark this answers is not in the document]
Comment (camilamacedo86) [reply pkg/controller/mobilesecurityservice/validations.go]: I agree. The idea here was impl after with can be required then the If will be required. I think we should leave in this way and do the other checks after.

File: pkg/controller/mobilesecurityserviceapp/controller.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (camilamacedo86) [reply pkg/controller/mobilesecurityserviceapp/controller.go]: In this case, it is the name of the MMS CR which should never be changed. 
They do not want to expose this kind of config in the APP CR. 
However, I added a const at least for it.

File: pkg/controller/mobilesecurityserviceapp/controller.go [hunk 7]
```diff
@@ -133,25 +133,42 @@ func (r *ReconcileMobileSecurityServiceApp) Reconcile(request reconcile.Request)
...
 		operatorNamespace = utils.OPERATOR_NAMESPACE_FOR_LOCAL_ENV
 	}
 
-	if err := r.client.Get(context.TODO(), types.NamespacedName{Name: "mobile-security-service", Namespace: operatorNamespace}, mssInstance); err != nil {
-		// Return and don't create
-		reqLogger.Info("Mobile Security Service instance resource not found. Mobile Security Service Application is required to create the application")
-		return reconcile.Result{}, nil
-	}
+	// GET MSS CR
+	r.client.Get(context.TODO(), types.NamespacedName{Name: "mobile-security-service", Namespace: operatorNamespace}, mssInstance)
```
[the remark this answers is not in the document]
Comment (camilamacedo86) [reply pkg/controller/mobilesecurityserviceapp/controller.go]: In this case, it is the name of the MMS CR which should never be changed.
They do not want to expose this kind of config in the APP CR.
However, I added a const at least for it.

File: pkg/controller/mobilesecurityservice/validations.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (camilamacedo86) [reply pkg/controller/mobilesecurityservice/validations.go]: Add a test for it.

File: pkg/controller/mobilesecurityservice/validations.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (camilamacedo86) [reply pkg/controller/mobilesecurityservice/validations.go]: The initial idea was it check all mandatory values. However, I am tracking a task for we think better over it. Maybe the best approach is not to ask to add the values in the CR and use const to define the default value and keep all as optional. It will also require changes in the readme.

File: pkg/models/sdk_config.go [hunk 8]
```diff
@@ -1,32 +1,11 @@
...
-	Version   string             `json:"version"`
-	Name      string             `json:"name"`
-	Namespace string             `json:"namespace"`
-	ClientID  string             `json:"clientId"`
-	Services  []SDKConfigService `json:"services,omitempty"`
-}
-
-func NewSDKConfig(m *mobilesecurityservicev1alpha1.MobileSecurityServiceApp, serviceUrl string) *SDKConfig {
-
-	cfg := new(SDKConfig)
```
Comment (craicoverflow) [inline pkg/models/sdk_config.go]: Constructing a model like this is not a common practice in Go, and you are taking multiple steps to build the `SDKConfig` struct, instead of 1.

`new(x)` sets each value in the struct to its default value. Next you are overwriting each value individually.

It would be better to define your constructor like so:

```go
func NewSDKConfig(m *mobilesecurityservicev1alpha1.MobileSecurityServiceApp, serviceUrl string) *SDKConfig {
	return &SDKConfig{
		ID:            m.Spec.AppId,
		Name:          m.Spec.AppName,
		Type:          "security",
		URL:           serviceUrl,
		ConfigService: *NewConfigService(serviceUrl),
	}
}
```

File: pkg/controller/mobilesecurityserviceapp/controller.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (laurafitzgerald) [reply pkg/controller/mobilesecurityserviceapp/controller.go]: @camilamacedo86 yes this was one of the improvements that I was required after my pr was merged last friday.

[the remark this answers is not in the document]
Comment (camilamacedo86) [reply pkg/controller/mobilesecurityserviceapp/controller.go]: I don't think we can expose this value @laurafitzgerald.

File: deploy/operator.yaml [hunk 9]
```diff
@@ -16,7 +16,7 @@ spec:
       containers:
         - name: mobile-security-service-operator
           # Replace this with the built image name
-          image: quay.io/aerogear/mobile-security-service-operator:master
+          image: docker.io/cmacedo/mobile-security-service-operator:SDK-CONFIG
```
Comment (austincunningham) [inline deploy/operator.yaml]: think this needs to change back :)

Comment (camilamacedo86) [reply deploy/operator.yaml]: Done

**Comment by davidffrench** [comment]
> Hi @davidffrench,
> 
> Done as your last comment. Note that you select the field which says that the JSON should have the type then after that you wrote that note.
> 
> Screenshot 2019-05-23 at 10 32 22
> 
> Screenshot 2019-05-23 at 10 32 15

That looks great to me @camilamacedo86 , thank you

**Comment by craicoverflow** [comment]
Successfully completed the verification steps.

Screenshot from 2019-05-23 15-05-20

File: pkg/controller/mobilesecurityservice/validations.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (craicoverflow) [reply pkg/controller/mobilesecurityservice/validations.go]: But this is what we have now and we don't even know if more checks will ever be added to this. So by leaving the current implementation is kind of premature optimization/over architecting the function for something that may never be needed. _If_ we need to change it in the future we can, but for now I don't think we need to.

File: pkg/controller/mobilesecurityservice/controller.go [hunk 10]
```diff
@@ -120,17 +120,17 @@ func (r *ReconcileMobileSecurityService) create(instance *mobilesecurityservicev
 func (r *ReconcileMobileSecurityService) buildFactory(reqLogger logr.Logger, instance *mobilesecurityservicev1alpha1.MobileSecurityService, resource string) runtime.Object {
 	reqLogger.Info("Check "+resource, "into the namespace", instance.Namespace)
 	switch resource {
-	case CONFIGMAP:
+	case ConfigMap:
 		return r.buildConfigMap(instance)
-	case DEPLOYMENT:
+	case Deployment:
```
Comment (craicoverflow) [inline pkg/controller/mobilesecurityservice/controller.go]: These changes should be in a separate pull request, as they are unrelated and make it harder to review the changes described in the scope.

Comment (craicoverflow) [reply pkg/controller/mobilesecurityservice/controller.go]: It is fine this time, but keep in mind for future :smile:

File: pkg/controller/mobilesecurityservice/helpers.go [hunk 11]
```diff
@@ -67,23 +68,11 @@ func getOAuthArgsMap(m *mobilesecurityservicev1alpha1.MobileSecurityService) []s
...
-}
-
-const letterBytes = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ"
-
 // RandStringBytes will return a string of n random bytes
 func RandStringBytes(n int) string {
 	b := make([]byte, n)
 	for i := range b {
-		b[i] = letterBytes[rand.Intn(len(letterBytes))]
+		b[i] = LetterBytes[rand.Intn(len(LetterBytes))]
```
Comment (craicoverflow) [inline pkg/controller/mobilesecurityservice/helpers.go]: This could have been left as it originally was, since currently the variable is not being used outside of this package.

File: pkg/controller/mobilesecurityservice/validations.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (craicoverflow) [reply pkg/controller/mobilesecurityservice/validations.go]: Thanks. I know it is not a new function but it was moved from a different package.

File: pkg/controller/mobilesecurityserviceapp/controller.go [hunk 12]
```diff
@@ -133,25 +133,42 @@ func (r *ReconcileMobileSecurityServiceApp) Reconcile(request reconcile.Request)
...
 
 	//Get the REST Service Endpoint
 	serviceAPI := service.GetServiceAPIURL(mssInstance)
 
 	//Check if the APP CR was marked to be deleted
 	isAppMarkedToBeDeleted := instance.GetDeletionTimestamp() != nil
 	if isAppMarkedToBeDeleted {
+
+		// If the Service was deleted and/or marked to be deleted
+		if err := r.client.Get(context.TODO(), types.NamespacedName{Name: utils.MobileSecurityServiceCRName, Namespace: operatorNamespace}, mssInstance); err != nil || mssInstance.GetDeletionTimestamp() != nil {
```
Comment (craicoverflow) [inline pkg/controller/mobilesecurityserviceapp/controller.go]: I could be wrong but the logic that occurs inside this condition is just removing the finalizers. So you could add a condition to only update if there are finalizers to delete. 

```suggestion
if err := r.client.Get(context.TODO(), types.NamespacedName{Name: utils.MobileSecurityServiceCRName, Namespace: operatorNamespace}, mssInstance); (err != nil || mssInstance.GetDeletionTimestamp() != nil) && len(instance.GetFinalizers()) > 0 {
```

File: pkg/models/sdk_config.go [hunk 13]
```diff
@@ -1,32 +1,11 @@
...
 
-//return the Service data for the SDK ConfigMap
-func getServices(m *mobilesecurityservicev1alpha1.MobileSecurityServiceApp, serviceUrl string) []SDKConfigService {
-	service := *NewSDKConfigServices(m, serviceUrl)
-	res := []SDKConfigService{}
-	res = append(res, service)
-	return res
+func NewSDKConfig(serviceUrl string) *SDKConfig {
+	return &SDKConfig{
+		URL:           serviceUrl,
```
Comment (craicoverflow) [inline pkg/models/sdk_config.go]: The spacing is very far apart here. Make sure you run `make fmt` before merging.
